### DISCUSSION TITLE
Is inline type-checking cached?

### DISCUSSION THREAD

**Discussion by isidentical** [geldata/gel#1876] [Q&A] [answered]
I was building an optimizer to take an autogenerated query of `SELECT Something FILTER .value = (SELECT Otherthing FILTER .value = '1')` into `SELECT Something FILTER .value[IS Otherthing].value = '1'` (of course the queries a bit more extensive than that; a simple real world [example](https://gist.github.com/isidentical/b1f90553510c8e31351655f8fb087754) Query 1 becomes Query 2 after the 'optimization' phase).

From what I can observe with very 'simple' equipment (just a `timeit.timeit` call with 5 iterations, the first query of my real example took `0.16152558857575058` (average, in total of 0.807627942878753). The second, and the 'optimized' (appereantly I'm doing something really wrong) took `3.687206262582913` (average, in total of 18.436031312914565 for 5 queries). 

Here is the benchmark script: [benchmark_raw_edgeql.py](https://github.com/reizio/reiz.io/blob/master/scripts/benchmark_raw_edgeql.py).

Any idea that how did I ended up with slowing my queries %1900 down with inlining them. I feel like this might be somewhat related with `[IS xxx]` controls, are they cached?

**Comment by elprans** [comment, marked as answer]
The `[IS ..]` variant being slower looks like a compilation pathology.  With respect to "optimization", your original query isn't less optimal than the (properly functioning) `[IS ]` variant, because in either case you are doing a `JOIN` with `Otherthing` in the current implementation.  This is because links don't contain type information, only the target object does.

Please open an issue with a releavant schema portion and we'll look into `[IS ..]` query performance there.
